vulkan: only use M-sized matmul on Apple GPUs

slp opened the pull request:
While Metal is the preferred backend while targeting Apple GPUs, using the Vulkan backend is very useful when you're indirectly accessing the GPU from a different context, such as when running in a Virtual Machine.

Right now, Vulkan backend's L-sized and S-sized matmuls are broken on Apple GPUs, so let's force using M-size with this vendor.

cc/ @0cc4m

slp committed:
vulkan: refactor guess_matmul_pipeline for vendor

Refactor ggml_vk_guess_matmul_pipeline to simplify adding per-vendor
conditionals.

Signed-off-by: Sergio Lopez <slp@redhat.com>

0cc4m commented:
It would be better to figure out why the other configurations are failing. What's your warp size? It writes it next to the device name.

cebtenzzre commented:
> What's your warp size? It writes it next to the device name.

The SIMD group size is always 32 on Metal, so I'd assume it's the same with Apple GPUs on Vulkan.

slp commented:
> It would be better to figure out why the other configurations are failing. What's your warp size? It writes it next to the device name.

I agree, this is intended as a temporary fix. I'll be exploring a proper fix and possible optimizations (the performance is roughly half of what you'd get with Metal) next, but that's going to take weeks (months?) and it'd be nice to have it working meanwhile.

> > What's your warp size? It writes it next to the device name.

> The SIMD group size is always 32 on Metal, so I'd assume it's the same with Apple GPUs on Vulkan.

Indeed, it's 32. Please also note there's a translation layer (MoltenVK) transforming Vulkan SPIR-V into MSL, so that could have a significant impact.

0cc4m commented:
> > It would be better to figure out why the other configurations are failing. What's your warp size? It writes it next to the device name.
> 
> I agree, this is intended as a temporary fix. I'll be exploring a proper fix and possible optimizations (the performance is roughly half of what you'd get with Metal) next, but that's going to take weeks (months?) and it'd be nice to have it working meanwhile.

Fair enough.

> > > What's your warp size? It writes it next to the device name.
> 
> > The SIMD group size is always 32 on Metal, so I'd assume it's the same with Apple GPUs on Vulkan.
> 
> Indeed, it's 32. Please also note there's a translation layer (MoltenVK) transforming Vulkan SPIR-V into MSL, so that could have a significant impact.

Then it's not this, there must be another issue with the other two shaders.

0cc4m commented on ggml-vulkan.cpp:
```diff
@@ -2032,12 +2033,50 @@ static uint32_t ggml_vk_guess_matmul_pipeline_align(ggml_backend_vk_context * ct
+        return ggml_vk_guess_matmul_pipeline_intel(ctx, bit16_x, bit16_y, aligned);
+    } else if (ctx->device.lock()->vendor_id == VK_VENDOR_ID_APPLE) {
+        return ggml_vk_guess_matmul_pipeline_apple(ctx, bit16_x, bit16_y, aligned);
+    }
```
Best use a switch statement here. There's bound to be more of these differences in the future, for example for Qualcomm.

slp replied:
Done.

0cc4m commented on ggml-vulkan.cpp:
```diff
@@ -2055,7 +2094,7 @@ static vk_pipeline* ggml_vk_guess_matmul_pipeline(ggml_backend_vk_context * ctx,
         return aligned ? &ctx->pipeline_matmul_f16_aligned_l : &ctx->pipeline_matmul_f16_l;
```
If you want to detangle this function completely, technically the `subgroupsize == 64` check is a check for AMD, so I guess that could go into a separate function as well.

slp replied:
Can we check for `vendor_id == VK_VENDOR_ID_AMD` there or is it better to keep checking for `subgroupsize == 64`?

0cc4m replied:
I think it's fine to check for AMD there, for now. I'm not aware of other GPUs with warps of size 64.

slp replied:
Done too.

0cc4m reviewed (commented):
Have you also checked the influence of k-splitting on Apple? I assume that worked fine? Can you post the full output of `GGML_VULKAN_RUN_TESTS=1`?

slp commented:
> Have you also checked the influence of k-splitting on Apple? I assume that worked fine? Can you post the full output of `GGML_VULKAN_RUN_TESTS=1`?

Sure, here is it: https://gist.github.com/slp/0a0c45c9cc3d490553e4d50472a2b4ad

On macOS Vulkan support is provided by MoltenVK, which translates Vulkan shaders into MSL. It's still too early to tell, but I suspect there's a mismatch in the translation of workgroup semantics to threadgroup/simdgroup. Here's the translation of the MULMAT shader to MSL: https://gist.github.com/slp/42bf29394f1a4c4c785cace7406e7b00

0cc4m commented:
> > Have you also checked the influence of k-splitting on Apple? I assume that worked fine? Can you post the full output of `GGML_VULKAN_RUN_TESTS=1`?
> 
> Sure, here is it: https://gist.github.com/slp/0a0c45c9cc3d490553e4d50472a2b4ad
> 
> On macOS Vulkan support is provided by MoltenVK, which translates Vulkan shaders into MSL. It's still too early to tell, but I suspect there's a mismatch in the translation of workgroup semantics to threadgroup/simdgroup. Here's the translation of the MULMAT shader to MSL: https://gist.github.com/slp/42bf29394f1a4c4c785cace7406e7b00

Interesting. Could you add the `vulkaninfo` output? Looks like some incompatibility in the translation, yeah. We'll find it eventually.

slp committed:
vulkan: only use M-sized matmul on Apple GPUs

L-sized and S-sized matmuls are broken on Apple GPUs, force using
M-size with this vendor.

Signed-off-by: Sergio Lopez <slp@redhat.com>

slp commented:
> > > Have you also checked the influence of k-splitting on Apple? I assume that worked fine? Can you post the full output of `GGML_VULKAN_RUN_TESTS=1`?
> > 
> > 
> > Sure, here is it: https://gist.github.com/slp/0a0c45c9cc3d490553e4d50472a2b4ad
> > On macOS Vulkan support is provided by MoltenVK, which translates Vulkan shaders into MSL. It's still too early to tell, but I suspect there's a mismatch in the translation of workgroup semantics to threadgroup/simdgroup. Here's the translation of the MULMAT shader to MSL: https://gist.github.com/slp/42bf29394f1a4c4c785cace7406e7b00
> 
> Interesting. Could you add the `vulkaninfo` output? Looks like some incompatibility in the translation, yeah. We'll find it eventually.

Here it goes: https://gist.github.com/slp/7e227bdb1eca24c05f4f79b4d1f8e3a6
